[HUDI-9123][RFC-91] Add RFC for storage based lock provider using conditional writes. (#12927)
apache/hudi: rfc/rfc-91/rfc-91.md

  the License.  You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->
# RFC-91: Storage-based lock provider using conditional writes

## Proposers

- @alexr17

## Approvers

 - @yihua
 - @danny0405

## Status

JIRA: [HUDI-9122](https://issues.apache.org/jira/browse/HUDI-9122)

## Abstract

Currently in Hudi, distributed locking relies on external systems like Zookeeper, which add complexity and extra dependencies. This RFC introduces a storage-based implementation of the `LockProvider` interface that utilizes conditional writes in cloud storage platforms (such as GCS and AWS S3) to implement a native distributed locking mechanism for Hudi. By directly integrating lock management with cloud storage, this solution reduces operational overhead, and ensures robust coordination during concurrent writes.

## Background

> AWS S3 recently introduced conditional writes, and GCS and Azure storage already support them. This RFC leverages these features to implement a distributed lock provider for Hudi using a leader election algorithm. In this approach, each process attempts an atomic conditional write to a file calculated using the table base path. The first process to succeed is elected leader and takes charge of exclusive operations. This method provides a straightforward, reliable locking mechanism without the need for external lock providers.

Reviewer: Let's also mention that there's limitation of existing implementation in `FileSystemBasedLockProvider` (see https://github.com/apache/hudi/pull/6071#discussion_r1061067820, https://github.com/apache/hudi/pull/7440/files#r1061068482) and conditional writes of the file system / storage are required for the storage-based lock provider to operate properly.

> AWS S3 recently introduced conditional writes, and GCS and Azure storage already support them. This RFC leverages these features to implement a distributed lock provider for Hudi using a leader election algorithm. In this approach, each process attempts an atomic conditional write to a file calculated using the table base path. The first process to succeed is elected leader and takes charge of exclusive operations. This method provides a straightforward, reliable locking mechanism without the need for external lock providers.

Reviewer: ```suggestion
AWS S3 recently introduced conditional writes, and GCS and Azure storage already support them. This RFC leverages these features to implement a distributed lock provider for Hudi using a leader election algorithm. In this approach, each process attempts an atomic conditional write to a file calculated using the table base path. The first process to succeed is elected leader and takes charge of exclusive operations. This method provides a straightforward, reliable locking mechanism without the need for additional distributed system.
```

## Implementation

> This design implements a leader election algorithm for Apache Hudi using a single lock file per table stored in cloud storage. Each table’s lock is represented by a JSON file with the following fields:

Reviewer 2: So the design mainly focuses on cloud storage, how about the HDFS which also has a large user base.

Author: Wow, looks like they just added java support: https://github.com/apache/opendal/pull/5664

So this might be doable. I think it still needs to be a pluggable implementation, in case OpenDAL does not support conditional writes in a storage type that someone wants to add a lock provider for.

Reviewer 2: > I think it still needs to be a pluggable implementation

+1 as of now.

Author: For HDFS and OpenDAL we will track them here:
- https://issues.apache.org/jira/browse/HUDI-9161
- https://issues.apache.org/jira/browse/HUDI-9162

Since both of these implementations are not available yet we will proceed with current S3/GCS client library approach to unblock Flink for 1.0.2 release, but can hopefully refactor in the future. cc @yihua

> This design implements a leader election algorithm for Apache Hudi using a single lock file per table stored in .hoodie folder. Each table’s lock is represented by a JSON file with the following fields:

Reviewer: ```suggestion
This design implements a leader election algorithm for Apache Hudi using a single lock file per table stored in .hoodie folder by default. Each table’s lock is represented by a JSON file with the following fields:
```

- owner: A unique UUID identifying the lock provider instance.
- expiration: A UTC timestamp indicating when the lock expires.
- expired: A boolean flag marking the lock as released.

> Example lock file path: `s3://bucket/table/.hoodie/.locks/table_lock.json`

Reviewer: ```suggestion
Example lock file path: `s3://bucket/table/.hoodie/.locks/table_lock.json`.  An advanced user might configure the lock file path in a separate location outside the bath path, but this is not recommended for proper concurrency control among multiple writers. 
```

Each `LockProvider` must implement `tryLock()` and `unlock()` however we also need to do our own lock renewal, therefore this implementation also has `renewLock()`. The implementation will import a service using reflection which writes to S3/GCS/Azure based on the provided location to write the locks. This ensures the main logic for conditional writes is shared regardless of the underlying storage.

> `tryLock()`

Reviewer: ```suggestion
`tryLock()`: guarantees that only one process can acquire the lock using the conditional write
```

- No Existing Lock: If the lock file doesn’t exist, a new lock file is created with the current instance’s details using a conditional write that only succeeds if the file is absent.
- Existing Lock – Not Expired: If a valid (non-expired) lock exists, the process refrains from taking the lock.

> - Existing Lock – Expired: If the lock file exists but is expired, a new lock file is conditionally written. This write uses a precondition based on the current file’s unique tag from cloud storage to ensure that no other process has updated it in the meantime.

Reviewer 3: > a new lock file is conditionally written

"new lock file" do you mean a brand new lock file will be created, or just update the current lock file with new lock payload? If it's the former, how is the "Existing lock" file dealt with.

Author: This is vague, I will change it. We are updating the current lock file with new lock payload. Essentially it overwrites the existing lock file.

> `renewLock()`
> - Purpose: Periodically extend the lock’s expiration (the heartbeat).

Reviewer: ```suggestion
`renewLock()`:  periodically extends the lock’s expiration (the heartbeat) to continue holding the lock if allowed.
```

- Mechanism: Update the lock file’s expiration using a conditional write that verifies the unique tag from the current lock state. If the tag does not match, the renewal fails, indicating that the lock has been lost.

`unlock()`

> - Purpose: Safely release the lock.

Reviewer 3: `unlock` would not delete lock file? so once the lock file is created in the first time lock, it would be always there, and will just be updated with conditional write?

Author: That's correct. Deleting the lock file is an unnecessary operation, we can just constantly update it.

> - Mechanism: Update the lock file’s expiration using a conditional write that verifies the unique tag from the current lock state. If the tag does not match, the renewal fails, indicating that the lock has been lost.

Reviewer: ```suggestion
- Update the lock file’s expiration using a conditional write that verifies the unique tag from the current lock state. If the tag does not match, the renewal fails, indicating that the lock has been lost.
```

> `unlock()`
> - Purpose: Safely release the lock.

Reviewer: ```suggestion
`unlock()`: safely releases the lock.
```

> - Mechanism: Update the existing lock file to mark it as expired. This update is performed with a conditional write that ensures the operation is only executed if the file’s unique tag still matches the one held by the lock owner. We do not delete the current lock file, this is an unnecessary operation.

Reviewer: ```suggestion
- Update the existing lock file to mark it as expired. This update is performed with a conditional write that ensures the operation is only executed if the file’s unique tag still matches the one held by the lock owner. We do not delete the current lock file, this is an unnecessary operation.
```

### Heartbeat Manager

Once a lock is acquired, a dedicated heartbeat task periodically calls renewLock() (typically every 30 seconds) to extend the expiration. This ensures the lock remains valid as long as the owning process (thread) is active. The heartbeat manager oversees this process, ensuring no other updates occur concurrently on the lock file. Each lock provider has one heartbeat manager with a single executor thread.

### Edge cases
- If the thread which acquired the lock dies, we stop the heartbeat.
- If the renewal fails past the expiration, we log an error, and stop the heartbeat. Other Hudi lock provider implementations are susceptible to this behavior. If a writer somehow loses access to Zookeeper, there is no way to tell the writer to exit gracefully.

> - If we are unable to start the heartbeat (renewal) we throw exception.

Reviewer 3: Will the lock holder perceive the exception in this case, or continue to finish the writing/txn

Author: If we cannot start the heartbeat we throw an exception but also release the lock. So the writer will never start a txn.

For renewal failures the writer will continue to finishing the txn

> - Clock drift: we allow for a maximum of 500ms of clock drift between nodes. A requirement of this lock provider is that all writers competing for the same lock must be writing from the same cloud provider (AWS/Azure/GCP).

Reviewer: Is this `500ms of clock drift` configurable?

Author: No, I don't think it should be either.

### New Hudi configs

> - `hoodie.write.lock.conditional_write.locks_location`: required, tells us where to write the locks files to.

Reviewer: ```suggestion
- `hoodie.write.lock.conditional_write.locks_location`: default empty String (indicating that "<table_base_path>/.hoodie/.locks/table_lock.json" is used as the lock file), tells us where to write the lock file to.
```

- https://cloud.google.com/storage/docs/metadata#generation-number

GCS has ETags, but they also have generation numbers, which are even better, and work for more use cases. Our current implementation already uses them, so they do not need further validation.

When we create the new lock file in tryLock we will use generationMatch(0). From GCP docs:
- *Passing the if_generation_match parameter to a method which retrieves a blob resource (e.g., Blob.reload) or modifies the blob (e.g., Blob.update) makes the operation conditional on whether the blob’s current generation matches the given value. As a special case, passing 0 as the value for if_generation_match makes the operation succeed only if there are no live versions of the blob.*

We can use the same logic for preconditions with overwrite operations using the currently stored lock file's generation number.

## Rollout/Adoption Plan

 - What impact (if any) will there be on existing users?
   - None 
 - If we are changing behavior how will we phase out the older behavior?
   - N/A
 - If we need special migration tools, describe them here.
   - N/A
 - When will we remove the existing behavior
   - N/A

## Test Plan

> We can write normal junit tests using testcontainers with GCS and S3 to simulate edge cases and general contention. Further adhoc testing will include the following scenarios:

Reviewer: Should we add a stress test purely on the new lock provider, i.e., creating a large number of concurrent threads, e.g., 1000, with each trying to lock and unblock with a random wait time of 0-1 second in between and validating that only one thread should hold the lock at any given time to make sure the implementation works as expected under extreme parallelism?

Author: yes this will be added, and updating test plan to reflect

> ## High-Frequency Commit and Table Service Test

Reviewer: ```suggestion
### High-Frequency Commit and Table Service Test
```

Run a long-running streaming ingestion process that continuously performs inserts, updates, and deletes. Ensure that frequent commits occur while table services like compaction and clustering operate concurrently. This test will help verify that the lock provider can handle overlapping operations without causing excessive delays or lock contention.

> ## Concurrent SQL and Spark Operations Test

Reviewer: Same for all subsections below to fold under `## Test Plan`
```suggestion
### Concurrent SQL and Spark Operations Test
```
